### PR TITLE
feat: better explain signing

### PR DESCRIPTION
**Pull request by weichweich**
fixes https://github.com/KILTprotocol/ticket/issues/2324

Actually show how signing is done using a keypair.

## Checklist:

- [ ] I have verified that the code works
- [ ] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [ ] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [ ] I have documented the changes (where applicable)

### CODE REVIEW

File: code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import * as Kilt from '@kiltprotocol/sdk-js'
+
+export async function signDidExtrinsic(
```
Comment (Dudleyneedham) [inline code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts]: Since it doesn't do much, it feels a little weird to wrap a `signDidExtrinsic`. The concern people have is how do you get these parameters.

File: code_examples/sdk_examples/src/core_features/did/09_did_signature.ts [hunk 2]
```diff
@@ -1,20 +1,32 @@
...
+import { u8aToHex } from '@polkadot/util'
+
+type KeyLookup = (parameter: {
+  didUri: Kilt.DidUri
+  keyRelationship: Kilt.VerificationKeyRelationship
+}) => Promise<{
+  key: Kilt.KiltKeyringPair
+  keyType: Kilt.VerificationKeyType
+  keyUri: Kilt.DidResourceUri
+}>
```
Comment (Dudleyneedham) [inline code_examples/sdk_examples/src/core_features/did/09_did_signature.ts:11]: This is a big leap. Where do we explain the concept of key storage. Most places we generate keys on the mnemonic. This has taken a big step in assuming that they store their key. Also, this would need explaining and in the text I do not see this.

File: code_examples/sdk_examples/src/core_features/utils/getRandomCtype.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import * as Kilt from '@kiltprotocol/sdk-js'
+
+export function getRandomCType(): Kilt.ICType {
+  // Random factor ensures that each created CType is unique and does not already exist on chain.
+  const randomFactor = Kilt.Utils.UUID.generate()
+  return Kilt.CType.fromProperties(`CType ${randomFactor}`, {
```
Comment (Dudleyneedham) [inline code_examples/sdk_examples/src/core_features/utils/getRandomCtype.ts]: Do you always want a random cType? What is the benefit of this?

It feels like bad behaviour to teach people.

It would be more practical to teach people to check if a CType is on-chain and reuse it.

File: code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts (hunk 1 above)
Comment (Dudleyneedham) [reply code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts]: I understand you inform them briefly about what the Params are, however, the issue is if people know what an extrinsic is.

File: code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  submitterAccount: Kilt.KiltKeyringPair,
+  fullDid: Kilt.DidUri,
+  signCallback: Kilt.SignExtrinsicCallback,
+  extrinsic: Kilt.SubmittableExtrinsic
+): Promise<void> {
+  // This results in a DID-signed tx that can be signed and submitted to
+  // the KILT blockchain by the account authorized in this operation (the `submitterAccount`).
```
Comment (Dudleyneedham) [inline code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts:13]: This feels more necessary in the description of the params for the extrinsic.

File: docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+title: Build DID Extrinsics
+---
+
+import TsJsBlock from '@site/src/components/TsJsBlock';
+
+import FullDidSignTx from '!!raw-loader!@site/code_examples/sdk_examples/src/core_features/did/08_full_did_sign_tx.ts';
+import FullDidBatch from '!!raw-loader!@site/code_examples/sdk_examples/src/core_features/did/08_full_did_batch.ts';
+
+A DID can be used to sign extrinsics.
+But not every extrinsics can be sign using a DID.
```
Comment (Dudleyneedham) [inline docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md]:
```suggestion
But not every extrinsic can be signed using a DID.
```

File: docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+id: full-did-batch
+title: Build DID Extrinsics
+---
+
+import TsJsBlock from '@site/src/components/TsJsBlock';
+
+import FullDidSignTx from '!!raw-loader!@site/code_examples/sdk_examples/src/core_features/did/08_full_did_sign_tx.ts';
+import FullDidBatch from '!!raw-loader!@site/code_examples/sdk_examples/src/core_features/did/08_full_did_batch.ts';
+
+A DID can be used to sign extrinsics.
```
Comment (Dudleyneedham) [inline docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md]:
```suggestion
DID keys can be used to sign extrinsic.
```

File: docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+But not every extrinsics can be sign using a DID.
+The Spiritnet blockchain offers two flavours of extrinsics.
+
+The first flavour can only be called using an account.
+These extrinsic require fee payment and need therefore be send from an account with sufficient balance.
```
Comment (Dudleyneedham) [inline docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md]:
```suggestion
These extrinsic require fee payment and need therefore be sent from an account with sufficient balance.
```

File: docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+The second flavour is callable from a DID.
```
Comment (Dudleyneedham) [inline docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md]:
```suggestion
The second callable flavour is from a DID.
```

File: docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+Since every extrinsic requires fees to be paid, this flavour needs to be wrapped inside a normal account extrinsic.
+The account will pay all fees and deposit that will be required.
```
Comment (Dudleyneedham) [inline docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md]:
```suggestion
All required fees and deposits will be paid by the account.
```

**Review by Dudleyneedham — CHANGES_REQUESTED** [review]
Nice PR. Good to see the documentation train a rolling.

My concern is that it still makes board assumptions about prior knowledge.

File: code_examples/sdk_examples/src/core_features/did/index.ts [hunk 10]
```diff
@@ -15,10 +15,13 @@ import { reclaimFullDidDeposit } from './11_full_did_deposit_reclaim'
 import { updateFullDid } from './07_full_did_update'
 
 import generateDidKeypairs from '../utils/generateKeypairs'
+import { getRandomCType } from '../utils/getRandomCtype'
+import { signDidExtrinsic } from './08_full_did_tx'
```
Comment (ntn-x2) [inline code_examples/sdk_examples/src/core_features/did/index.ts]: Can you move this back up with the other imports from the same folder?

File: code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts (hunk 1 above)
Comment (ntn-x2) [inline code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts]: This functio not only signs but also submits, so it should be renamed.

File: docs/develop/01_sdk/02_cookbook/01_dids/04_full_did_query.md [hunk 11]
```diff
@@ -1,6 +1,6 @@
 ---
 id: did-query
-title: Resolve a DID
+title: Resolve a Full DID
```
Comment (ntn-x2) [inline docs/develop/01_sdk/02_cookbook/01_dids/04_full_did_query.md]: Why this change? The API is the same for both.

File: docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md (hunk 7 above)
Comment (ntn-x2) [reply docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md]: It sounds like the second flavour of extrinsics do not involve any fee payment. I would probably switch them around, mentioning that there are some operations that require a DID signature. Everything else is the same. Both involve fee payment and regular account signatures.

File: docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+This document describes how to sign the DID extrinsics.
+The KILT SDK provides two functions for signing DID extrinsics.
+The first function signs a single extrinsic while the second one batches multiple extrinsics together.
+
+## Single extrinsics
+
+To build a single extrinsic, you need to provide:
```
Comment (ntn-x2) [inline docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md]:
```suggestion
To sign a single extrinsic, you need to provide:
```

File: docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+* the DID that should be the origin of the extrinsic
```
Comment (ntn-x2) [inline docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md]:
```suggestion
* the DID that wants to sign the extrinsic (also called *origin* of the extrinsic)
```

File: docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+* a [`SignCallback`](../06_signCallback.md) that signs the extrinsic
+* the extrinsic that should be submitted
```
Comment (ntn-x2) [inline docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md]:
```suggestion
* the extrinsic that should be signed and submitted
```

File: code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts (hunk 1 above)
Comment (ntn-x2) [reply code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts]: What about having these functions return the extrinsic object?

File: docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md (hunk 13 above)
Comment (ntn-x2) [reply docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md]: Maybe the order of these elements can match the order of the parameters in the function signature? That would make it a bit easier for the users to follow the two at the same time.

**Review by ntn-x2 — COMMENTED** [review]
I like the improvement in clarity, althought there are some rough edges I would like smoothed out. Please also check that the Amplify goes through successfully, as the preview does not match the latest version of the PR>

File: docs/develop/01_sdk/02_cookbook/01_dids/04_full_did_query.md (hunk 11 above)
Comment (weichweich) [reply docs/develop/01_sdk/02_cookbook/01_dids/04_full_did_query.md]: Since the file is called `full_did_query` i assumed it was about full DIDs. If it's about light and full i would rename the file and change the title.

Comment (ntn-x2) [reply docs/develop/01_sdk/02_cookbook/01_dids/04_full_did_query.md]: Yes, I would say it's better to have a "resolve a DID" everywhere, as users should not really know what the difference is.

Comment (weichweich) [reply docs/develop/01_sdk/02_cookbook/01_dids/04_full_did_query.md]: [1dfe4c3](https://github.com/KILTprotocol/docs/pull/183/commits/1dfe4c30677266b1921b675d98b6c43606e27959)

File: code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts (hunk 1 above)
Comment (weichweich) [reply code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts]: Most of these things should be separate guides tho.

* "the extrinsic that should be signed and submitted" if they don't have this, they don't need the guide. The extrinsic to submit will be something they find in other guides. and there we would link here. It doesn't make sense to explain where this comes from.
* SignCallback - We link to a guide on how to build this
* "the DID that wants to sign the extrinsic (also called origin of the extrinsic)" we should link to the DID creation guide.
* "and the address of the account that pays for the fees." here we should link to the concept guide that explains the difference between the DID and account and maybe how to create an account.

But explaining all this also in this guide would be too much

File: code_examples/sdk_examples/src/core_features/did/09_did_signature.ts (hunk 2 above)
Comment (weichweich) [reply code_examples/sdk_examples/src/core_features/did/09_did_signature.ts:11]: Where to get the keys from is a new topic. This is far to complex and depends on quite a lot of things. We can't explain this in this section. That's the reason for having this lookup method. They need to look it up themself.

File: code_examples/sdk_examples/src/core_features/utils/getRandomCtype.ts (hunk 3 above)
Comment (weichweich) [reply code_examples/sdk_examples/src/core_features/utils/getRandomCtype.ts]: Yes we shouldn't mention that it's a random ctype. We use this call since it's the easiest to construct. but we can use any. So i removed all mentions of the ctype in the visible code [babcf11](https://github.com/KILTprotocol/docs/pull/183/commits/babcf117a6541e2ec8fbe3e2e23248b254915005)

File: docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md (hunk 7 above)
Comment (weichweich) [reply docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md]: But we say in the second flavor that these need to be wrapped using the first flavor inorder for fees to be paid. So i think it's clear that both need to be paid.

Also we don't have terminology for these two types. We should give them names so that we can distinguish them. Account Extrinsic and DID Extrinsic?

Comment (ntn-x2) [reply docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md]: At least I had this doubt when going over the text, I think at least the text can be reworded to explain the concept a bit differently.
I am ok with the Account vs DID Extrinsic distinction.

Comment (weichweich) [reply docs/develop/01_sdk/02_cookbook/01_dids/06_full_did_tx.md]: I tried to make it clearer. [e16cf85](https://github.com/KILTprotocol/docs/pull/183/commits/e16cf85b28773e1eb24a03ee8b74d28ee4a30b4c)

File: code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts (hunk 1 above)
Comment (weichweich) [reply code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts]: I would at least sign them using an account. This makes it clear that it's a two step process.

File: code_examples/sdk_examples/src/core_features/did/index.ts (hunk 10 above)
Comment (weichweich) [reply code_examples/sdk_examples/src/core_features/did/index.ts]: [5e9eb88](https://github.com/KILTprotocol/docs/pull/183/commits/5e9eb88fdaf23a924bff8b16512fb7bffe5623fa)

File: code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts (hunk 1 above)
Comment (weichweich) [reply code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts]: [55eefba](https://github.com/KILTprotocol/docs/pull/183/commits/55eefbae2f0a781b498b72e71481d55342609509)

File: code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts [hunk 15]
```diff
@@ -0,0 +1,23 @@
+import * as Kilt from '@kiltprotocol/sdk-js'
+// Just a helper to get an extrinsic
+import getExtrinsic from '../utils/getExtrinsic'
+
+export async function signAndSendDidExtrinsic(
```
Comment (ntn-x2) [inline code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts]: What about `signAndSubmit`? It's a term more people are familiar with.

File: code_examples/sdk_examples/src/core_features/did/08_full_did_batch.ts [hunk 16]
```diff
@@ -1,17 +1,6 @@
...
-    },
-    age: {
-      type: 'integer'
-    }
-  })
-}
+// Just a helper to get an extrinsic
+import getExtrinsic from '../utils/getExtrinsic'
 
 export async function batchCTypeCreationExtrinsics(
```
Comment (ntn-x2) [inline code_examples/sdk_examples/src/core_features/did/08_full_did_batch.ts]: Also this name is misleading. What does `batch` mean? What about `signAndSubmitCTypeCreationBatch`?

**Review by ntn-x2 — COMMENTED** [review]
just a couple of suggestions about the naming of the helper functions, otherwise LGTM 😊

File: code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts (hunk 15 above)
Comment (weichweich) [reply code_examples/sdk_examples/src/core_features/did/08_full_did_tx.ts]: good point we should stick to submit. [76db79a](https://github.com/KILTprotocol/docs/pull/183/commits/76db79a5f6a52895fbd0591a82334b2736439905)

File: code_examples/sdk_examples/src/core_features/did/08_full_did_batch.ts (hunk 16 above)
Comment (weichweich) [reply code_examples/sdk_examples/src/core_features/did/08_full_did_batch.ts]: good catch. We actually don't want to mention ctypes anymore. I renamed it to `signAndSubmitDidExtrinsicBatch`. So it also matches the other function. [ee24291](https://github.com/KILTprotocol/docs/pull/183/commits/ee24291b1852d2887938265dfcfb9efb876b0e48)

**Review by ntn-x2 — APPROVED** [review]
